spec/src/main/asciidoc/servlet-spec-body.adoc

```diff
@@ -1303,6 +1303,139 @@ the header value to an `int`, a `NumberFormatException` is thrown. If
+  * Any `"."` or `".."` segment with any encoded characters
+  * Any `".."` segment preceded by an empty segment
+  * The `"\"` character encoded or not.
+  * Any control characters either encoded or not.
+
+==== Example URIs
+> TODO more work needed on this section. Perhaps some parts need to be moved to other sections.
+
+Consider the following URIs with a forbidden security constraint at `/secret/*` and a servlet registered at /dispatch/* that forwards request to the string return from pathInfo()
+
+. Example URIs
+|===
+| Encoded URI path  | Decoded Path | Rejected  
+
+| /public/file.txt | /public/file.txt |
+| /public//file.txt | /public/file.txt |
+| /public/;/file.txt | /public/file.txt |
+| /PUBLIC/file.txt | /PUBLIC/file.txt |
+| /public%2Ffile.txt | /public%2Ffile.txt | 400 
+| /public%5Cfile.txt | /public\file.txt | 400
+| /public%00/file.txt | /public␀/file.txt | 400?
+| /public/./file.txt | /public/file.txt | 
+| /public/.;/file.txt | /public/file.txt | 400
+| /public/%2e/file.txt | /public/file.txt | 400
+| /public/%2e;/file.txt | /public/file.txt | 400
+| /../docroot/public/file.txt | /../docroot/public/file.txt | 400
+| /public/dir/../file.txt | /public/file.txt |  
+| /public//../file.txt | /file.txt | 400?
+| /public/;param/../file.txt | /file.txt | 400?
+| /public/dir/..;/file.txt | /public/file.txt | 400
+| /public/dir/%2e%2e/file.txt | /public/file.txt | 400
+| /public/dir/%2e%2e;/file.txt | /public/file.txt | 400
+| /WEB-INF/web.xml | /WEB-INF/web.xml | 404 or 403?
+| /web-inf/web.xml | /web-inf/web.xml | 404 or 403?
+| /WEB-IN~1.DIR/web.xml | ? | ?
+| /WEB-INF;/web.xml | /WEB-INF/web.xml | 404
+| /WEB-INF%2Fweb.xml | /WEB-INF%2Fweb.xml| 400
+| /WEB-INF%5Cweb.xml | /WEB-INF\web.xml | 400 
+| /WEB-INF%00/web.xml | /WEB-INF␀/web.xml | 400?
+| /WEB-INF/./web.xml | /WEB-INF/web.xml | 404
+| /public/../WEB-INF/web.xml | /WEB-INF/web.xml | 404 or 403?
+| /public/..;/WEB-INF/web.xml | /WEB-INF/web.xml | 404 or 403?
+| /public/%2e%2e/WEB-INF/web.xml | /WEB-INF/web.xml | 404 or 403?
+| /public/%2e%2e;/WEB-INF/web.xml | /WEB-INF/web.xml | 404 or 403?
+| /secret/private.xml | /secret/private.xml | 403
+| /SeCreT/private.xml | /SeCreT/private.xml | TODO move to another section an 404 due to non-canonical
+| /SECRET~1.DIR/private.xml  |  | TODO move to another section an 404 due to non-canonical
+| /secret;/private.xml | | 403 
+| /secret%2Fprivate.xml | /secret%2Fprivate.xml | 400
+| /secret%5Cprivate.xml | /secret\private.xml | 400
+| /secret%00/private.xml | /secret␀/private.xml | 400? 
+| /./secret/private.xml | /secret/private.xml| 403
+| /.;/secret/private.xml | /secret/private.xml| 403
+| /%2e/secret/private.xml | /secret/private.xml | 400
+| /%2e;/secret/private.xml | /secret/private.xml  | 400
+| /public/../secret/private.xml | /secret/private.xml | 403
+| /public/..;/secret/private.xml | /secret/private.xml | 403
+| /public/%2e%2e/secret/private.xml | /secret/private.xml | 400
```

Comment:
@stuartwdouglas said:
> So turns out I actually hate both the options when it comes to /%2e%2e/ being allowed :-)
> As written where we have different behavior to the RFC if you have a path like:
> `/public/%2e%2e/private/secret.txt`
> The end result will be:
> `/private/secret.txt` (which is different to what we had discussed before, where we were going to map to the context first). This is a security issue if a front end proxy is securing /private
> If we do it the other way around and we end up with:
> `/public/../private/secret.txt`
> This will map to /public as expected, but you end up with a path that starts with a .. segment, which these rules are largely supposed to prevent.
> I guess this is ok because %2e%2e is disallowed by default, but things seem to get very problematic when it is allowed.

Reply:
I tend to agree.
I think that if `/%2e%2e/` segments are allowed then, by default, we should throw if and of the existing path methods are used and thus they will only be available from `getRequestUri()` or new API.

If we did this, then we could decode after dot-segment removal and be much closer to the RFC.